packages/peregrine/lib/store/middleware/auth.js

```diff
@@ -0,0 +1,60 @@
+import BrowserPersistence from '../../util/simplePersistence';
+import userActions, { signOut } from '../actions/user';
+
+const timers = new Map();
+const storage = new BrowserPersistence();
+const SET_TOKEN = userActions.setToken.toString();
+const CLEAR_TOKEN = userActions.clearToken.toString();
+const GET_DETAILS = userActions.getDetails.request.toString();
+
+const isSigningIn = type => type === SET_TOKEN || type === GET_DETAILS;
+const isSigningOut = type => type === CLEAR_TOKEN;
+
+/**
+ * This function adheres to Redux's middleware pattern.
+ *
+ * @param {Store} store The store to augment.
+ * @returns {Function}
+ */
+const scheduleSignOut = store => next => action => {
+    const { dispatch } = store;
+
+    if (isSigningIn(action.type)) {
```

Comment:
So when we sign in, we create a timer to remove the token based on the TTL of the token just set? So this is similar to what [getItem](https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/util/simplePersistence.js#L33-L45) does but is also able to dispatch the `signOut` action and clear the token from redux (persisted as `M2_VENIA_BROWSER_PERSISTENCE__signin_token`).

The current iteration of the auth link also uses the [`getItem` mechanism](https://github.com/magento/pwa-studio/blob/develop/packages/venia-concept/src/index.js#L33) which should delete the token if expired.

Both seem to handle similar situations related to client-side invalidation of the token, but I still don't know if we gracefully handle a user token that invalidates server-side before the client-side expiry.

Reply:
Correct, `getItem` can't dispatch `signOut`, so the Redux store contains stale state (`isSignedIn: true`) when there's an expired token or no token in storage.

This PR doesn't handle a server-side invalidation, but I believe those are going to be much less frequent. Expiration should be the 99% case.

Reply:
>  This PR doesn't handle a server-side invalidation, but I believe those are going to be much less frequent. Scheduled expiration should be the 99% case.

Right, Tommy's solution gets us there, assuming the error messages are correct. That said, I don't know enough about auth invalidation so I'll trust that you when you say that a timeout expiry is the most common scenario.